File: internal/adr/adr_test.go
package adr_test

import (
	"fmt"
	"os"
	"path/filepath"
	"testing"

	"github.com/delightfulhammers/telesis/internal/adr"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func setupADRDir(t *testing.T) string {
	t.Helper()
	dir := t.TempDir()
	require.NoError(t, os.MkdirAll(filepath.Join(dir, "docs", "adr"), 0o755))
	return dir
}

func TestCreateFirstADR(t *testing.T) {
	rootDir := setupADRDir(t)

	path, err := adr.Create(rootDir, "use-cobra")
	require.NoError(t, err)

	assert.Equal(t, filepath.Join(rootDir, "docs", "adr", "ADR-001-use-cobra.md"), path)
	assert.FileExists(t, path)

	content, err := os.ReadFile(path)
	require.NoError(t, err)
	assert.Contains(t, string(content), "# ADR-001: use-cobra")
	assert.Contains(t, string(content), "## Status")
	assert.Contains(t, string(content), "Proposed")
	assert.Contains(t, string(content), "## Context")
	assert.Contains(t, string(content), "## Decision")
	assert.Contains(t, string(content), "## Consequences")
}

func TestCreateSequentialADRs(t *testing.T) {
	rootDir := setupADRDir(t)

	path1, err := adr.Create(rootDir, "first")
	require.NoError(t, err)
	assert.Contains(t, path1, "ADR-001-first.md")

	path2, err := adr.Create(rootDir, "second")
	require.NoError(t, err)
	assert.Contains(t, path2, "ADR-002-second.md")

	path3, err := adr.Create(rootDir, "third")
	require.NoError(t, err)
	assert.Contains(t, path3, "ADR-003-third.md")
}

func TestCreateADRWithExistingGap(t *testing.T) {
	rootDir := setupADRDir(t)
	adrDir := filepath.Join(rootDir, "docs", "adr")

	// Create ADR-005 directly (simulating a gap)
	require.NoError(t, os.WriteFile(
		filepath.Join(adrDir, "ADR-005-existing.md"),
		[]byte("# ADR-005: existing\n"),
		0o644,
	))

	path, err := adr.Create(rootDir, "next")
	require.NoError(t, err)
	assert.Contains(t, path, "ADR-006-next.md")
}

func TestCreateADRRejectsEmptySlug(t *testing.T) {
	rootDir := setupADRDir(t)

	_, err := adr.Create(rootDir, "")
	assert.Error(t, err)
	assert.Contains(t, err.Error(), "slug")
}

func TestCreateADRRejectsInvalidSlug(t *testing.T) {
	tests := []struct {
		name string
		slug string
	}{
		{"spaces", "has spaces"},
		{"uppercase", "HasUpperCase"},
		{"special chars", "has/slashes"},
		{"dots", "has.dots"},
		{"starts with hyphen", "-leading"},
		{"ends with hyphen", "trailing-"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			rootDir := setupADRDir(t)
			_, err := adr.Create(rootDir, tt.slug)
			assert.Error(t, err)
			assert.Contains(t, err.Error(), "slug")
		})
	}
}

func TestCreateADRMissingDirectory(t *testing.T) {
	rootDir := t.TempDir()
	// No docs/adr/ directory

	_, err := adr.Create(rootDir, "something")
	assert.Error(t, err)
}

func TestNextNumberEmptyDir(t *testing.T) {
	rootDir := setupADRDir(t)

	num, err := adr.NextNumber(filepath.Join(rootDir, "docs", "adr"))
	require.NoError(t, err)
	assert.Equal(t, 1, num)
}

func TestNextNumberWithExisting(t *testing.T) {
	rootDir := setupADRDir(t)
	adrDir := filepath.Join(rootDir, "docs", "adr")

	for i := 1; i <= 3; i++ {
		require.NoError(t, os.WriteFile(
			filepath.Join(adrDir, fmt.Sprintf("ADR-%03d-test.md", i)),
			[]byte(fmt.Sprintf("# ADR-%03d: test\n", i)),
			0o644,
		))
	}

	num, err := adr.NextNumber(adrDir)
	require.NoError(t, err)
	assert.Equal(t, 4, num)
}

func TestNextNumberIgnoresNonADRFiles(t *testing.T) {
	rootDir := setupADRDir(t)
	adrDir := filepath.Join(rootDir, "docs", "adr")

	require.NoError(t, os.WriteFile(
		filepath.Join(adrDir, "README.md"),
		[]byte("# ADRs\n"),
		0o644,
	))

	num, err := adr.NextNumber(adrDir)
	require.NoError(t, err)
	assert.Equal(t, 1, num)
}

File: src/adr/adr.test.ts
import { describe, it, expect } from "vitest";
import { mkdirSync, writeFileSync, readFileSync, existsSync } from "node:fs";
import { join } from "node:path";
import { create, nextNumber } from "./adr.js";
import { useTempDir } from "../test-utils.js";

const makeTempDir = useTempDir("adr-test");

const setupADRDir = (): string => {
  const dir = makeTempDir();
  mkdirSync(join(dir, "docs", "adr"), { recursive: true });
  return dir;
};

describe("adr", () => {
  it("creates first ADR", () => {
    const rootDir = setupADRDir();
    const path = create(rootDir, "use-cobra");

    expect(path).toBe(join(rootDir, "docs", "adr", "ADR-001-use-cobra.md"));
    expect(existsSync(path)).toBe(true);

    const content = readFileSync(path, "utf-8");
    expect(content).toContain("# ADR-001: use-cobra");
    expect(content).toContain("## Status");
    expect(content).toContain("Proposed");
    expect(content).toContain("## Context");
    expect(content).toContain("## Decision");
    expect(content).toContain("## Consequences");
  });

  it("creates sequential ADRs", () => {
    const rootDir = setupADRDir();

    const path1 = create(rootDir, "first");
    expect(path1).toContain("ADR-001-first.md");

    const path2 = create(rootDir, "second");
    expect(path2).toContain("ADR-002-second.md");

    const path3 = create(rootDir, "third");
    expect(path3).toContain("ADR-003-third.md");
  });

  it("creates ADR with existing gap", () => {
    const rootDir = setupADRDir();
    const adrDir = join(rootDir, "docs", "adr");

    writeFileSync(join(adrDir, "ADR-005-existing.md"), "# ADR-005: existing\n");

    const path = create(rootDir, "next");
    expect(path).toContain("ADR-006-next.md");
  });

  it("rejects empty slug", () => {
    const rootDir = setupADRDir();
    expect(() => create(rootDir, "")).toThrow("slug");
  });

  it("rejects invalid slugs", () => {
    const invalidSlugs = [
      { name: "spaces", slug: "has spaces" },
      { name: "uppercase", slug: "HasUpperCase" },
      { name: "special chars", slug: "has/slashes" },
      { name: "dots", slug: "has.dots" },
      { name: "starts with hyphen", slug: "-leading" },
      { name: "ends with hyphen", slug: "trailing-" },
    ];

    for (const { slug } of invalidSlugs) {
      const rootDir = setupADRDir();
      expect(() => create(rootDir, slug)).toThrow("slug");
    }
  });

  it("fails with missing directory", () => {
    const rootDir = makeTempDir();
    expect(() => create(rootDir, "something")).toThrow();
  });

  describe("nextNumber", () => {
    it("returns 1 for empty directory", () => {
      const rootDir = setupADRDir();
      const num = nextNumber(join(rootDir, "docs", "adr"));
      expect(num).toBe(1);
    });

    it("returns next with existing files", () => {
      const rootDir = setupADRDir();
      const adrDir = join(rootDir, "docs", "adr");

      for (let i = 1; i <= 3; i++) {
        writeFileSync(
          join(adrDir, `ADR-${String(i).padStart(3, "0")}-test.md`),
          `# ADR-${String(i).padStart(3, "0")}: test\n`,
        );
      }

      const num = nextNumber(adrDir);
      expect(num).toBe(4);
    });

    it("ignores non-ADR files", () => {
      const rootDir = setupADRDir();
      const adrDir = join(rootDir, "docs", "adr");

      writeFileSync(join(adrDir, "README.md"), "# ADRs\n");

      const num = nextNumber(adrDir);
      expect(num).toBe(1);
    });
  });
});
